### ISSUE TITLE
FileUpload: implement validateContentType

### ISSUE THREAD

**Issue by tandraschko** [primefaces/primefaces#4244]
Currently we just validate the filename.
With Files#probeContentType we can get the real content type and validate it based on the "accept" attribute.
The problem is that #probeContentType is not reliable enough and the validation is only secure when apache tika is included as dependency:
`<dependency>
  <groupId>org.apache.tika</groupId>
  <artifactId>tika-java7</artifactId>
  <version>1.19.1</version>
</dependency>`

discussion: https://github.com/primefaces/primefaces/issues/2791
PR: https://github.com/primefaces/primefaces/pull/4242

**Comment by tandraschko** [comment]
@cnsgithub it would also be great if we could make the validate method bit smaller, it's quite complex.
Maybe splitup into validateFileName and validateFileContent? I'm you sure you will find a good way :D

**Comment by cnsgithub** [comment]
> The problem is that #probeContentType is not reliable enough and the validation is only secure when apache tika is included as dependency:
> <dependency> <groupId>org.apache.tika</groupId> <artifactId>tika-core</artifactId> <version>1.19.1</version> </dependency>

Not sure if `tika-core` includes `tika-java7` as well which contains the service provider implementation?

**Comment by tandraschko** [comment]
me neither, would be great if you can check it.

**Comment by cnsgithub** [comment]
Nope, `tika-core `doesn't work, we need `tika-java7 `(which includes tika-core, but not the other way around):
```xml
        <dependency>
            <groupId>org.apache.tika</groupId>
            <artifactId>tika-java7</artifactId>
            <version>1.19.1</version>
            <scope>test</scope>
        </dependency>
```

**Comment by tandraschko** [comment]
Good that tika-java7 is optional, the dependecies are ~50mb ;)

**Comment by tandraschko** [comment]
it's actually a no-go for such a small feature.

**Comment by cnsgithub** [comment]
I agree, that's really big. That's because tika is not only able to recognize content types by some magic bytes, but also supports extraction of meta data from various file formats...

Edit: Maybe excluding tika-parsers may decrease the size...

**Comment by cnsgithub** [comment]
Just as a side note, it would be appropriate, if servlet containers would make use of frameworks like Apache Tika to properly prefill `javax.servlet.http.Part#getContentType` for example...

**Comment by tandraschko** [comment]
tika-parsers is the same size or nearly the same.
tika-core isn't enough, right?

**Comment by cnsgithub** [comment]
> tika-parsers is the same size or nearly the same.

I thought WITHOUT tika-parsers. This is probably the biggest module.

> tika-core isn't enough, right?

That's right.

**Comment by tandraschko** [comment]
ahh :D 
ok, plz try :P

**Comment by cnsgithub** [comment]
Seems to work with this configuration, but cannot guarantee that detection is as reliable as before:

```xml
        <dependency>
            <groupId>org.apache.tika</groupId>
            <artifactId>tika-java7</artifactId>
            <version>1.19.1</version>
            <scope>test</scope>
            <exclusions>
                <exclusion>
                    <groupId>org.apache.tika</groupId>
                    <artifactId>tika-parsers</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
```

**Comment by tandraschko** [comment]
could we verify this via unittests?

**Comment by cnsgithub** [comment]
I've already provided a bunch of tests, mostly using magic numbers. But that's of course in no way exhaustive.

**Comment by cnsgithub** [comment]
Maybe you could ask Apache folks about that? :P

**Comment by tandraschko** [comment]
why me? :D i don't even use fileupload ;)

**Comment by cnsgithub** [comment]
Because you are the member of Apache Software Foundation, not me! ;)

**Comment by tandraschko** [comment]
everyone can ask questions on the dev mailing list :P

**Comment by cnsgithub** [comment]
Actually I would leave the decision to `p:fileUpload `users if they are willing to include hundreds of thousands of bytes or not. Users might even have there own SPI implementations plugged in. So I think it's not worth the effort from the PF side. I would simply recommend to include `tika-java7 `in the user guide and that's it.

Edit: Maybe my answer would have been different if Apache Tika had its issue tracking on github... ;)

**Comment by cnsgithub** [comment]
> @cnsgithub it would also be great if we could make the validate method bit smaller, it's quite complex.

Done, PR: https://github.com/primefaces/primefaces/pull/4249

**Comment by fr4x1nu5** [comment]
> everyone can ask questions on the dev mailing list :P

Just posted to dev@tika.apache.org.

**Comment by cnsgithub** [comment]
> everyone can ask questions on the dev mailing list :P

Just posted to dev@tika.apache.org, permalink: https://lists.apache.org/thread.html/a944008199d4c6d0f47c5ddcb8b1041672072eb835ae526db546f2d3@%3Cdev.tika.apache.org%3E

> Hey guys,
> 
> we recently contributed some security improvements to the popular JSF component library PrimeFaces to validate specified accepted content types of uploaded files at server side. Therefore we make use of Java's Files.probeContentType which automatically picks up registered java.nio.file.spi.FileTypeDetector service implementations. The default implementations however primarily check the file extension only by doing registry lookups or something like that. That's insufficient from a security point of view. That's why we will recommend to have Apache Tika in the classpath, more strictly speaking the tika-java7 dependency.
> 
> Now there are two questions regarding the use of TikaFileTypeDetector and the required dependencies:
> 1. Transitive dependencies of tika-java7 are really big (more than 50 megabyte). I know that Apache Tika is not just about file type detection but very much more like meta data extraction that we don't need at all. Would it be okay to exclude tika-parsers which makes up the biggest module without losing file type detection abilities?
> 2. Unfortunately, TikaFileTypeDetector defaults to perform most efficiently by having included a short circuit if the content type can be guessed from the file extension. This is however insecure since we want to protect our users from tampered file uploads. We always want to use deep (and expensive) content type analysis by looking at the magic bytes or something like that. We currently work around this limitation by explicitly putting .tmp as the file name's extension to have Tika detected application/octet-stream and force it to go ahead. But that's some kind of white box knowledge to rely on which is not that good. Can you somehow provide different implementations of java.nio.file.spi.FileTypeDetector, say "eager" and "lazy"? Please note that we are not allowed to introduce required dependencies, i.e. using Tika directly is not an option.
> 
> Here are the related issues: https://github.com/primefaces/primefaces/issues/2791 and https://github.com/primefaces/primefaces/issues/4244
> And the pull requests already merged into PrimeFaces 6.3: https://github.com/primefaces/primefaces/pull/4242 and https://github.com/primefaces/primefaces/pull/4249
> 
> Thanks for your response in advance
> 
> Kind regards, cnsgithub

**Comment by cnsgithub** [comment]
Response from @tballison

> Sorry for the late reply!
> show/hide original text
> 
> >Would it be okay to exclude tika-parsers which makes up the biggest
> module without losing file type detection abilities?
> 
> It _should work_ for straight mime/magic detection. However, you need
> some of the items within tika-parsers to figure out specific POIFS
> (OLE/ooxml) and package types...e.g. what type of an ooxml or OLE file
> is it, really, etc. These are the key classes that you'd be missing
> without tika-parsers:
> org.apache.tika.parser.microsoft.POIFSContainerDetector
> org.apache.tika.parser.pkg.ZipContainerDetector And, you'd also need
> to include POI and commons-compress as dependencies. In
> short...experiment with those file types and see if the results are ok
> for you.
> show/hide original text
> 
> >But that's some kind of white box knowledge to rely on which is not that good. Can
> you somehow provide different implementations of
> java.nio.file.spi.FileTypeDetector, say "eager" and "lazy"?
> 
> I wouldn't object to this. To confirm, we'd have two implementations
> in the java7 package and you'd do something like this in your code:
> Class.forName("org.apache.tika.filetypedetector.TikaBytesFileTypeDetector");
> 
> Open a ticket on our JIRA, and I'll see what I can do.
> 
> Fellow devs, any other recommendations?
> 
> Best,
> 
> Tim
> 

What do you guys think?
